torchaudio/transforms.py

```diff
@@ -557,12 +550,104 @@ def forward(self, waveform: Tensor) -> Tensor:
         else:
             mel_specgram = self.amplitude_to_DB(mel_specgram)
 
-        # (..., channel, n_mels, time).transpose(...) dot (n_mels, n_mfcc)
-        # -> (..., channel, time, n_mfcc).transpose(...)
-        mfcc = torch.matmul(mel_specgram.transpose(-2, -1), self.dct_mat).transpose(-2, -1)
+        # (..., channel, n_mels, time) dot (n_mels, n_mfcc) -> (..., channel, n_nfcc, time)
+        mfcc = torch.einsum("...cnt,nl->...clt", mel_specgram, self.dct_mat)
```

Comment:
@nateanl What is the performance implication of `einsum`?

Reply:
The current version of ``einsum`` should be faster than matmul. Check: https://github.com/pytorch/pytorch/issues/32591#issuecomment-856936116

Reply:
I thought so as well, but for a simple [test script](https://gist.github.com/Joool/6c3b24cdbb9a915bf7943714f4d46b23) (uses white noise as audio), the old implementation does outperform `einsum` on my machine:

```
PyTorch version 1.10.0.dev20210617+cu111
Torchaudio version 0.10.0a0+bac32ec

CPU:
Old implementation: 0.2768233451060951
Einsum implementation: 0.33643896970897913

GPU:
Old implementation: 0.23426104336977005
Einsum implementation: 0.30553871812298894

```

Reply:
The results are similar when I use the `torch.benchmark` ([script](https://gist.github.com/Joool/b7ed70e4c9b3c001473c398254b80687)):

```
PyTorch version 1.10.0.dev20210617+cu111
Torchaudio version 0.10.0a0+bac32ec
CPU:
     Old | 35.8   
     Einsum | 42.0

Times are in microseconds (us).

GPU:
      Old | 24
      Einsum | 31.6

Times are in microseconds (us).
```

Reply:
I have the same result by trying your script and change the ``duration`` to 5.
```
PyTorch version 1.9.0
Torchaudio version 0.10.0a0+0210807
CPU:
Old implementation: 5.004824341
Einsum implementation: 5.26490228
```
Based on that, we should change it back to the old version to get a more effective performance. 🙂

Reply:
Done in f07fb93. I additionally removed the unnecessary batch unpacking which was present in the old implementation.

Old:
https://github.com/pytorch/audio/blob/108a32d92cd77733841d6c61ab91f8e3ce995f35/torchaudio/transforms.py#L285-L294

New:
```
mel_specgram = torch.matmul(specgram.transpose(-1, -2), self.fb).transpose(-1, -2)
```

Which is slightly more efficient (I assume since you do not reshape as much):
```
PyTorch version 1.10.0.dev20210617+cu111
Torchaudio version 0.10.0a0+bac32ec
CPU:
    Old |                   35.8
    Einsum |                41.5
    Old simplified |        30.9 

Times are in microseconds (us).

GPU:
    Old |                   24
    Einsum |                30.3
    Old simplified |        18

Times are in microseconds (us).
```